### PR TITLE
Handing large polygons intersecting high-resolution rasters

### LINKED ISSUE #37: Memory constraints for large watersheds
So far I've been unable to run the watershed count function for the cities of `New Orleans | LA` and `Saint Louis | MO`. This is because the watersheds are so large that the raster cropping and mapping seems to require a huge amount of memory. Attempted with 2 entire nodes on PIC and failed, so we need a solution. Generally, the raster cropping and mapping is slow for larger watersheds, so anything to identify those cases and speed things up would be valuable.

### PR DESCRIPTION
**Pull request by crvernon**
Development addressed the need to handle memory and processing time constraints that prevented large polygons from being able to fully extract grid cell values from a high-resolution raster.  The following was conducted to deal with the issue:

- Large polygons for `New Orleans | LA` and `Saint Louis | MO` are now split into equal parts (the default is 40 though when this could likely be reduced depending upon the memory available in the machine processing the data)
- Each equal part is then used to extract the required information from the source raster in parallel (number of cores - 1).  
- The outputs are then combined and returned.

This PR fixes #37

### CODE REVIEW

**Comment by KristianNelson** [comment]
The package runs fine on my machine, as it will on yours I assume, but when I ran the check it returned this error: 

> checking package dependencies ... ERROR
  Namespace dependency not required: 'units'

Not sure where `units` is within the package to take the dependency out.

**Comment by KristianNelson** [comment]
@swd-turner @crvernon All checks have passed. Should be good to go after you both go through it.

**Comment by KristianNelson** [comment]
Ran the larger watersheds and the smaller watersheds. Everything's working smoothly now.

File: R/count_teleconnections.R [hunk 1]
```diff
@@ -6,19 +6,21 @@
 #' @param crop_file_path path of crop cover raster
 #' @param dams_file_path path of National Inventory of Dams "NID" point file
 #' @param cities a vector of cities to be included in the count. If omitted, all cities will be included.
+#' @param poly_slices integer for how may parts to split the watersheds polygons into to enable faster zonal stats
 #' @details counts teleconnections assoicated with water supply catchments associated with each city
```
Comment (swd-turner) [inline R/count_teleconnections.R]: param documentation for `n_cores` missing

File: R/io.R [hunk 2]
```diff
@@ -210,19 +210,20 @@ get_ucs_power_plants <- function(ucs_file_path,
...
 #'
-#' Get the matrix of raster values represented in the input raster dataset
-#' when restricted to the input watershed polygons for a target city.
+#' Get the frequency per class of raster values represented in the input raster dataset
+#' when restricted to the input watershed polygons for a target city.  This ignores NA.
 #'
 #' @param raster_object character. An object of class RasterLayer.
 #' @param polygon character. A polygon to define spatial boundary of raster value counts (e.g. a given city's watersheds)
-#' @return matrix of raster values in target polygons
+#' @return matrix of raster value and frequency in target polygons
```
Comment (swd-turner) [inline R/io.R]: no longer returns a matrix, right? It's a table of crop types present and their frequency of occurrence.

File: R/io.R [hunk 3]
```diff
@@ -234,12 +235,15 @@ get_raster_val_matrix <- function(raster_object, polygon) {
   polys <- polygon %>%
     st_transform(crs = r_crs)
 
-  # calculate the number of unique land classes from the input raster that are in the target polygons
+  # calculate the frequency of unique land classes from the input raster that are in the target polygons
   n_lcs <- crop(raster_object, polys) %>%
     mask(polys) %>%
-    raster::getValues()
+    freq(useNA = "no") %>%
+    as.data.frame()
```
Comment (swd-turner) [inline R/io.R]: Can we use `as_tibble` instead please, for consistency? The renaming can be done in the same pipe:

`... as_tibble() %>% rename(Group.1 = value, x = count)`

**Pull request merged**